src/types/datetime.rs

```diff
@@ -83,11 +90,17 @@ impl PyDate {
     ///
     /// This is equivalent to `datetime.date.fromtimestamp`
     pub fn from_timestamp(py: Python, timestamp: i64) -> PyResult<Py<PyDate>> {
-        let args = PyTuple::new(py, &[timestamp]);
+        let time_tuple = PyTuple::new(py, &[timestamp]);
 
         unsafe {
-            let ptr = (PyDateTimeAPI.Date_FromTimestamp)(PyDateTimeAPI.DateType, args.as_ptr());
-            Py::from_owned_ptr_or_err(py, ptr)
+            #[cfg(PyPy)]
+            let ptr = PyDate_FromTimestamp(time_tuple.as_ptr());
```

Comment:
Can you clarify why this is necessary? Does this mean that on PyPy the `PyDateTimeAPI` struct doesn't have a FromTimestamp method? Why is this?

Reply:
this is what PyPy has for (as time of writing this 7.0) for datetime:
![image](https://user-images.githubusercontent.com/2467993/54947604-622c8a00-4f43-11e9-8dac-471e3cb96538.png)

cpython has indeed some extra methods:
![image](https://user-images.githubusercontent.com/2467993/54947662-86886680-4f43-11e9-997d-2c7da85c275e.png)

So I've had to go around the current cpyext API and call those methods directly.

Reply:
@omerbenamram Can you use actual blocks of text instead of screen shots from now on?  You can still get code highlighting with blocks of 3 backticks.

As for why this is different, I'm a bit stumped. Maybe we can make a PR to PyPy.

The thing is that the capsule object *is* part of the public interface for PyO3, so ideally we want that to be a seamless experience for users who are using the `ffi` bindings directly.

Reply:
I'm not sure I get your last point about the capsule object being a public interface, from what I've understood (as a user of the library) - the public API is `PyDateTime` (https://github.com/PyO3/pyo3/blob/master/src/types/datetime.rs), which was where I've felt that I could bridge the differences.

This is an internal decision by PyPy, I'm not entirely sure why they chose to implement it this way 😕 But I don't really see a way to rewire these functions to behave the same.

Reply:
@omerbenamram I mean that the unsafe bindings are also part of the public interface, not just the safe Rust ones. If people are calling the `ffi::datetime::PyDateTimeAPI` functions directly, they'll have to do the compatibility shims for PyPy themselves, which is not good.

I think ideally we can make a PR to PyPy to add these to the C API. We may also be able to do something like populate our `PyDateTimeAPI` object with function pointers to the relevant `cpyext` functions, if necessary, though I think that might make the ABI less stable than it already is (not sure what guarantees are made about ABI compatibility here).

Reply:
>  If people are calling the ffi::datetime::PyDateTimeAPI functions directly, they'll have to do the compatibility shims for PyPy themselves, which is not good.

While I agree that it would be ideal if cpython and pypy exposed the same functions in their c api, I think that it's ok that users have to care for compatibility themselves when using `ffi` directly, as pyo3 is providing a safe and uniform abstraction.

Reply:
> While I agree that it would be ideal if cpython and pypy exposed the same functions in their c api, I think that it's ok that users have to care for compatibility themselves when using ffi directly, as pyo3 is providing a safe and uniform abstraction.

Yeah, that's probably true, but if we *can* provide a better interface, I think we should. One of the reasons `cpyext` even exists is so that pre-existing extension modules will just work without having to explicitly target pypy. If you're explicitly targeting pypy, my understanding is that a `cffi`-style approach is better than a C extension approach. This is why I think it would be best to abstract away the differences as much as possible.